rfcs/029-work_state_modelling/README.md

```diff
@@ -0,0 +1,177 @@
+    type ImageId = IdentifierState.Unidentified
+  }
+
+  case class Merged() extends WorkState {
+    type Id = IdentifierState.Unminted
+    type ImageId = IdentifierState.Unidentified
+  }
+
+  case class Denormalised(
+    relations: Relations[Denormalised],
+    sourceIdentifier: SourceIdentifier,
+    identifiedType: String = classOf[Identified].getSimpleName,
+  ) extends WorkState {
+    type Id = IdentifierState.Unminted
+    type ImageId = IdentifierState.Unidentified
+  }
+
+  case class Identified(
+    canonicalId: String,
+    relations: Relations[Identified],
+  ) extends WorkState {
+    type Id = IdentifierState.Minted
+    type ImageId = IdentifierState.Identified
+  }
+}
+
+/** IdentifierState represents the state of individual pieces of work data
+  * (such as genres, concepts etc), with Unminted used to indicate
+  * pre-minter state and Minted used to indicate post-minter state
+  */
+sealed trait IdentifierState
+
+object IdentifierState {
+
+  sealed trait Unminted extends IdentifierState
+  sealed trait Minted extends IdentifierState
+
+  case class Unidentified(
+    sourceIdentifier: SourceIdentifier
+  ) extends Unminted
+
+  case class Identified(
+    sourceIdentifier: SourceIdentifier,
+    canonicalId: String
+  ) extends Minted
+
+  case class Unidentifiable() extends Minted with Unminted
+}
+
+/** Work contains the work itself. It is parameterised by it's state, meaning
+  * the same type of Work can be in a number of possible states depending on
+  * where in the pipeline it is. This allows us to easily add new types of work
+  * (such as if Collection is decided to be a seperate type to StandardWork),
+  * with the state of the work in the pipeline being an orthogonal concern.
+  */
+sealed trait Work[State <: WorkState] {
+  val sourceIdentifier: SourceIdentifier
+  val version: Int
+}
+
+case class StandardWork[State <: WorkState](
+  sourceIdentifier: SourceIdentifier,
+  version: Int,
+  data: WorkData[State],
+  state: State,
+) extends Work[State]
+
+case class RedirectedWork[State <: WorkState](
+  sourceIdentifier: SourceIdentifier,
+  version: Int,
+  state: State,
+) extends Work[State]
+
+case class InvisibleWork[State <: WorkState](
+  sourceIdentifier: SourceIdentifier,
+  version: Int,
+  data: WorkData[State],
+  state: State,
+) extends Work[State]
+
+/** WorkData contains data common to all types of works that can exist at any
+  * stage of the pipeline. It is parameterised by a WorkState in order for
+  * individual pieces of data to have the correct IdentifierState depending on
+  * the stage of the pipeline.
+  */
+case class WorkData[State <: WorkState](
+  title: Option[String] = None,
+  genres: List[Genre[State#Id]] = Nil,
+  images: List[Image[State#ImageId, State#Id]] = Nil,
+  ...
+)
+
+/** Relations contain the related works for some particular work, and only exist
+  * at the Denormalised and Identified stages of the pipeline.
+  */
+case class Relations[State <: WorkState](
```

Comment:
How will we represent different kinds of relation (i.e. relationships within a collection may not be the only kind of relation we wish to represent).

Reply:
As far as I understand, this is intended for different sorts of relations beyond just collections, with @jtweed designing it with other use cases in mind. All relations which form a directed acyclic graph of works can be represented this way: we will need to make sure we prevent cycles, which I think should be fine in practice with collections and the other individual use cases being of tree form.

Reply:
How do we describe the type of relation though? Would that be another `ADT`? i.e. 

```scala
sealed trait Relationship {
   parts: Option[List[Work[State]]]
  // etc
}

case class CollectionRelationship(parts: Option[List[Work[State]]]) extends Relationship
```

Reply:
Maybe @jtweed can chime in but I think the idea is to make relations agnostic to what sort of thing they are in the source system (so from the perspective of the consumer of the API they do not know / care that it is an archive they are looking at, rather than a hand picked bunch of photos of a certain theme or something else).

Fwiw if I was to encode the relation type in the model I would probably do it more like this than using inheritance:

```scala
case class Relations[State](
  parts: Option[List[Relation[State]]],
  ...
)

case class Relation[State](
  relationType: RelationType,
  work: Work[State]
)

sealed trait RelationType

object RelationType {
  case object Collection extends RelationType
  ...
}
```

Reply:
```scala
case class Relation[State](
  relationType: RelationType,
  work: Work[State]
)
``` 
👍  

If we don't care about the kind of relationship then we shouldn't encode it. Are we certain we don't care @jtweed ?